checks/shell_download_validate.go

```diff
@@ -0,0 +1,628 @@
+		return true
+	})
+
+	return cmdValidated, nil
+}
+
+func isSuFetchStdinExecute(node syntax.Node, cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	ce, ok := node.(*syntax.CallExpr)
+	if !ok {
+		return false, nil
+	}
+
+	c, ok := extractCommand(ce)
+	if !ok {
+		return false, nil
+	}
+
+	if !isSuCommand(c) {
+		return false, nil
+	}
+
+	cs, ok := extractSuCommand(ce.Args)
+	if !ok {
+		return false, nil
+	}
+
+	// We have extracted the command. Now parse it.
+	in := strings.NewReader(cs)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, fmt.Errorf("syntax.NewParser: %w", err)
+	}
+
+	isFetch := false
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if isFetchStdinExecute(node, cmd, pathfn, logf) {
+			isFetch = true
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return isFetch, nil
+}
+
+func validateCommandIsNotSuFetchToStdinExecute(cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if err != nil {
+			return false
+		}
+		var b bool
+		b, err = isSuFetchStdinExecute(node, cmd, pathfn, logf)
+		if err != nil {
+			return false
+		} else if b {
+			cmdValidated = false
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return cmdValidated, err
+}
+
+func validateCommandIsNotFetchToStdinExecute(cmd, pathfn string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	r1, err := validateCommandIsNotNotSuFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return r1, err
+	}
+
+	r2, err := validateCommandIsNotSuFetchToStdinExecute(cmd, pathfn, logf)
+	if err != nil {
+		return r2, err
+	}
+
+	return r1 && r2, nil
+}
+
+func recordFetchFileFromString(cmd string) (fmap map[string]bool, ok bool, err error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return nil, false, ErrParsingShellCommand
+	}
+
+	files := make(map[string]bool)
+	syntax.Walk(f, func(node syntax.Node) bool {
+		if err != nil {
+			return false
+		}
+		var b bool
+		var fn string
+		fn, b, err = recordFetchFileFromNode(node)
+		if err != nil {
+			return false
+		} else if b {
+			files[fn] = true
+		}
+
+		// Continue walking the node graph.
+		return true
+	})
+
+	return files, true, err
+}
+
+func validateCommandIsNotFileExecute(cmd, pathfn string, files map[string]bool,
+	logf func(s string, f ...interface{})) (bool, error) {
+	in := strings.NewReader(cmd)
+	f, err := syntax.NewParser().Parse(in, "")
+	if err != nil {
+		return false, ErrParsingShellCommand
+	}
+
+	cmdValidated := true
+	syntax.Walk(f, func(node syntax.Node) bool {
+		// Check if we're calling a file we previously downloaded.
+		if isExecuteFiles(node, cmd, pathfn, files, logf) {
+			cmdValidated = false
```

Comment:
do we need to continue walking in this case?

Reply:
yes, as I want to provide all offending lines to the user. Otherwise they'll see a single error, fix it.. and scorecard will continue reporting errors.